src/libstd/collections/hash/map.rs

```diff
@@ -311,38 +318,49 @@ pub struct HashMap<K, V, S = RandomState> {
 fn search_hashed<K, V, M, F>(table: M,
                              hash: SafeHash,
                              mut is_match: F)
-                             -> SearchResult<K, V, M> where
+                             -> InternalEntry<K, V, M> where
     M: Deref<Target=RawTable<K, V>>,
     F: FnMut(&K) -> bool,
 {
+    // Worst case, we'll find one empty bucket among `size + 1` buckets.
     let size = table.size();
-    let mut probe = Bucket::new(table, hash);
+    let mut probe = if let Some(probe) = Bucket::new(table, hash) {
+        probe
+    } else {
+        return InternalEntry::TableIsEmpty;
+    };
     let ib = probe.index();
 
-    while probe.index() != ib + size {
-        let full = match probe.peek() {
-            Empty(b) => return TableRef(b.into_table()), // hit an empty bucket
-            Full(b) => b
+    loop {
+        let bucket = match probe.peek() {
+            Empty(empty) => {
+                // Found a hole!
+                return InternalEntry::Vacant(NoElem(empty));
+            }
+            Full(bucket) => bucket
         };
 
-        if full.distance() + ib < full.index() {
+        let robin_ib = bucket.index() as int - bucket.displacement() as int;
+
+        if (ib as isize) < robin_ib {
             // We can finish the search early if we hit any bucket
             // with a lower distance to initial bucket than we've probed.
-            return TableRef(full.into_table());
+            return InternalEntry::Vacant(NeqElem(bucket, robin_ib as usize));
         }
 
         // If the hash doesn't match, it can't be this one..
-        if hash == full.hash() {
+        if hash == *bucket.read().0 {
             // If the key doesn't match, it can't be this one..
-            if is_match(full.read().0) {
-                return FoundExisting(full);
+            if is_match(bucket.read().1) {
+                return InternalEntry::Occupied(OccupiedEntryState {
```

Comment:
I don't know what the changes in this function have to do with in-place hashmap growth.